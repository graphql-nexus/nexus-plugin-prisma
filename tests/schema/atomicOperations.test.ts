import { objectType } from 'nexus'
import { generateSchemaAndTypes } from '../__utils'

it('support atomic operations (default)', async () => {
  const datamodel = `
  model ModelName {
    id      Int @id @default(autoincrement())
    name    String
    amount  Int
  }
`
  const ModelName = objectType({
    name: 'ModelName',
    definition(t: any) {
      t.model.id()
    },
  })

  const Mutation = objectType({
    name: 'Mutation',
    definition(t: any) {
      t.crud.createOneModelName()
      t.crud.deleteOneModelName()
      t.crud.updateOneModelName()
      t.crud.upsertOneModelName()
      t.crud.updateManyModelName()
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [Mutation, ModelName])

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})

it('disable atomic operations', async () => {
  const datamodel = `
  model ModelName {
    id      Int @id @default(autoincrement())
    name    String
    amount  Int
  }
`
  const ModelName = objectType({
    name: 'ModelName',
    definition(t: any) {
      t.model.id()
    },
  })

  const Mutation = objectType({
    name: 'Mutation',
    definition(t: any) {
      t.crud.createOneModelName()
      t.crud.deleteOneModelName()
      t.crud.updateOneModelName()
      t.crud.upsertOneModelName()
      t.crud.updateManyModelName()
    },
  })

  const { schemaString: schema, typegen } = await generateSchemaAndTypes(datamodel, [Mutation, ModelName], {
    atomicOperations: false,
  })

  expect(schema).toMatchSnapshot('schema')
  expect(typegen).toMatchSnapshot('typegen')
})
