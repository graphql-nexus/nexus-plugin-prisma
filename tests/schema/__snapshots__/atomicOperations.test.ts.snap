// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`disable atomic operations: schema 1`] = `
"type Mutation {
  createOneModelName(data: ModelNameCreateInput!): ModelName!
  deleteOneModelName(where: ModelNameWhereUniqueInput!): ModelName
  updateOneModelName(data: ModelNameUpdateInput!, where: ModelNameWhereUniqueInput!): ModelName
  upsertOneModelName(where: ModelNameWhereUniqueInput!, create: ModelNameCreateInput!, update: ModelNameUpdateInput!): ModelName!
  updateManyModelName(data: ModelNameUpdateManyMutationInput!, where: ModelNameWhereInput): AffectedRowsOutput!
}

type ModelName {
  id: Int!
}

input ModelNameCreateInput {
  name: String!
  amount: Int!
}

input ModelNameWhereUniqueInput {
  id: Int
}

input ModelNameUpdateInput {
  name: String
  amount: Int
}

type AffectedRowsOutput {
  count: Int!
}

input ModelNameUpdateManyMutationInput {
  name: String
  amount: Int
}

input ModelNameWhereInput {
  AND: [ModelNameWhereInput!]
  OR: [ModelNameWhereInput!]
  NOT: [ModelNameWhereInput!]
  id: IntFilter
  name: StringFilter
  amount: IntFilter
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input StringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedStringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

type Query {
  ok: Boolean!
}"
`;

exports[`disable atomic operations: typegen 1`] = `
"import * as Typegen from '@kenchi/nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  ModelName: Prisma.ModelName
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    modelNames: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'amount'
      ordering: 'id' | 'name' | 'amount'
    }
  },
  ModelName: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    modelName: 'ModelName'
    modelNames: 'ModelName'
  },
  Mutation: {
    createOneModelName: 'ModelName'
    updateOneModelName: 'ModelName'
    updateManyModelName: 'AffectedRowsOutput'
    deleteOneModelName: 'ModelName'
    deleteManyModelName: 'AffectedRowsOutput'
    upsertOneModelName: 'ModelName'
  },
  ModelName: {
    id: 'Int'
    name: 'String'
    amount: 'Int'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  ModelName: Typegen.NexusPrismaFields<'ModelName'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;

exports[`support atomic operations (default): schema 1`] = `
"type Mutation {
  createOneModelName(data: ModelNameCreateInput!): ModelName!
  deleteOneModelName(where: ModelNameWhereUniqueInput!): ModelName
  updateOneModelName(data: ModelNameUpdateInput!, where: ModelNameWhereUniqueInput!): ModelName
  upsertOneModelName(where: ModelNameWhereUniqueInput!, create: ModelNameCreateInput!, update: ModelNameUpdateInput!): ModelName!
  updateManyModelName(data: ModelNameUpdateManyMutationInput!, where: ModelNameWhereInput): AffectedRowsOutput!
}

type ModelName {
  id: Int!
}

input ModelNameCreateInput {
  name: String!
  amount: Int!
}

input ModelNameWhereUniqueInput {
  id: Int
}

input ModelNameUpdateInput {
  name: StringFieldUpdateOperationsInput
  amount: IntFieldUpdateOperationsInput
}

type AffectedRowsOutput {
  count: Int!
}

input ModelNameUpdateManyMutationInput {
  name: StringFieldUpdateOperationsInput
  amount: IntFieldUpdateOperationsInput
}

input ModelNameWhereInput {
  AND: [ModelNameWhereInput!]
  OR: [ModelNameWhereInput!]
  NOT: [ModelNameWhereInput!]
  id: IntFilter
  name: StringFilter
  amount: IntFilter
}

input StringFieldUpdateOperationsInput {
  set: String
}

input IntFieldUpdateOperationsInput {
  set: Int
  increment: Int
  decrement: Int
  multiply: Int
  divide: Int
}

input IntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input StringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

input NestedIntFilter {
  equals: Int
  in: [Int!]
  notIn: [Int!]
  lt: Int
  lte: Int
  gt: Int
  gte: Int
  not: NestedIntFilter
}

input NestedStringFilter {
  equals: String
  in: [String!]
  notIn: [String!]
  lt: String
  lte: String
  gt: String
  gte: String
  contains: String
  startsWith: String
  endsWith: String
  not: NestedStringFilter
}

type Query {
  ok: Boolean!
}"
`;

exports[`support atomic operations (default): typegen 1`] = `
"import * as Typegen from '@kenchi/nexus-plugin-prisma/typegen'
import * as Prisma from '@prisma/client';

// Pagination type
type Pagination = {
    first?: boolean
    last?: boolean
    before?: boolean
    after?: boolean
}

// Prisma custom scalar names
type CustomScalars = 'No custom scalars are used in your Prisma Schema.'

// Prisma model type definitions
interface PrismaModels {
  ModelName: Prisma.ModelName
}

// Prisma input types metadata
interface NexusPrismaInputs {
  Query: {
    modelNames: {
      filtering: 'AND' | 'OR' | 'NOT' | 'id' | 'name' | 'amount'
      ordering: 'id' | 'name' | 'amount'
    }
  },
  ModelName: {

  }
}

// Prisma output types metadata
interface NexusPrismaOutputs {
  Query: {
    modelName: 'ModelName'
    modelNames: 'ModelName'
  },
  Mutation: {
    createOneModelName: 'ModelName'
    updateOneModelName: 'ModelName'
    updateManyModelName: 'AffectedRowsOutput'
    deleteOneModelName: 'ModelName'
    deleteManyModelName: 'AffectedRowsOutput'
    upsertOneModelName: 'ModelName'
  },
  ModelName: {
    id: 'Int'
    name: 'String'
    amount: 'Int'
  }
}

// Helper to gather all methods relative to a model
interface NexusPrismaMethods {
  ModelName: Typegen.NexusPrismaFields<'ModelName'>
  Query: Typegen.NexusPrismaFields<'Query'>
  Mutation: Typegen.NexusPrismaFields<'Mutation'>
}

interface NexusPrismaGenTypes {
  inputs: NexusPrismaInputs
  outputs: NexusPrismaOutputs
  methods: NexusPrismaMethods
  models: PrismaModels
  pagination: Pagination
  scalars: CustomScalars
}

declare global {
  interface NexusPrismaGen extends NexusPrismaGenTypes {}

  type NexusPrisma<
    TypeName extends string,
    ModelOrCrud extends 'model' | 'crud'
  > = Typegen.GetNexusPrisma<TypeName, ModelOrCrud>;
}
  "
`;
